### PR TITLE
refactor(pipeline/pii): Phase 6 — split PiiRedactor.ts into per-category modules + Section 13 ESLint cap

### PR DESCRIPTION
**Pull request by sergienko4**
# Phase 6 — Split `PiiRedactor.ts` into 13 focused per-category modules

## Summary

`src/Scrapers/Pipeline/Types/PiiRedactor.ts` was a **996-LoC** blob that owned every PII-redaction concern: classification, per-category redaction strategies (account / card / Israeli ID / phone / name / merchant / amount / token / OTP / cookie / URL / HTML), the JSON-body walker, the Pino censor, and the error-log helpers. This PR mechanically splits it into **13 focused per-category modules** + a `Types.ts` shared-types module + a `CommonHelpers.ts` grapheme-utility module + a `Facade.ts` dispatch table under `Types/PiiRedactor/`, leaving a **30-effective-LoC re-export shim** so every caller import + every existing test stays byte-identical.

Direct successor of Phase 5 ([#277](https://github.com/sergienko4/israeli-bank-scrapers/pull/277), `ScrapeAutoMapper.ts` 1637 → 27-eff shim). Same playbook, same per-commit husky-14-gate discipline, same `max-lines` guardrail at the end.

**No public API removal or rename.** Two new public additions land in commit 1 (`redact`, `REDACTION_ERROR_HINT`) plus two internal symbols promoted to public via the tombstone shim (`REDACTED_HINT`, `classifyKey`); every legacy export is preserved.

## Sub-module layout

| Folder                              | File                  | Eff LoC | Responsibility                                                                       |
|-------------------------------------|-----------------------|--------:|--------------------------------------------------------------------------------------|
| `PiiRedactor/`                      | `Types.ts`            |      28 | Shared types + sentinels (`REDACTED_HINT`, `REDACTION_ERROR_HINT`, brand types, JSON union, `isPiiRedactionDisabled`) |
| `PiiRedactor/`                      | `CommonHelpers.ts`    |      10 | `buildSegmenter` + `graphemeCount` Intl-grapheme helpers (length-class hints)        |
| `PiiRedactor/`                      | `Facade.ts`           |     165 | Strategy registry + `classify` / `redact` / `createCensorFn` dispatch + `PATH_TAIL_TO_CATEGORY` |
| `PiiRedactor/`                      | `Account.ts`          |      28 | Classify + redact account numbers (last-4 hint)                                      |
| `PiiRedactor/`                      | `Card.ts`             |      15 | Classify + redact card numbers (last-4 + BIN hint)                                   |
| `PiiRedactor/`                      | `IsraeliId.ts`        |      17 | Israeli-ID format detection + redaction                                              |
| `PiiRedactor/`                      | `Phone.ts`            |      17 | Phone-number format detection + redaction                                            |
| `PiiRedactor/`                      | `Name.ts`             |      10 | Human-name length-class redaction                                                    |
| `PiiRedactor/`                      | `Merchant.ts`         |      10 | Merchant length-class redaction                                                      |
| `PiiRedactor/`                      | `Amount.ts`           |      20 | Amount redaction (sign preserved)                                                    |
| `PiiRedactor/`                      | `Html.ts`             |      29 | HTML body redaction                                                                  |
| `PiiRedactor/`                      | `Url.ts`              |      76 | URL path/query redaction + `PII_QUERY_KEYS`                                          |
| `PiiRedactor/`                      | `JsonBody.ts`         |     134 | JSON walker dispatching per-key into the strategy registry                           |
| `PiiRedactor/`                      | `AuthCredentials.ts`  |      43 | `token` / `otp` / `cookie` — always-redact (no `PII_REDACTION=off` bypass)           |
| `PiiRedactor/`                      | `ErrorLog.ts`         |      21 | `redactErrorMessage` + `redactSensitiveEnum` + `SENSITIVE_SCRAPER_ENUMS` (CodeQL #28) |
| (shim)                              | `PiiRedactor.ts`      |  **30** | Pure re-export tombstone                                                             |

Largest new file is **165 effective LoC** (Facade — strategy registry + dispatch table); second-largest is **134 effective LoC** (JsonBody walker). Both well under the new 200-line ceiling. Thirteen of 15 modules ≤43 eff.

## Commit sequence (7 code commits + 1 post-ship chore, all green on husky's 14-gate pre-commit)

| # | SHA       | Subject                                                                                              |
|--:|-----------|------------------------------------------------------------------------------------------------------|
| 1 | `c28499fe` | refactor(pipeline/pii): [phase-6-commit-1] add unified redact() + Facade scaffolding + invariant tests |
| 2 | `1ebe6003` | refactor(pipeline/pii): split Identifier redactors (Account/Card/IsraeliId) into per-category modules |
| 3 | `8d722ab5` | refactor(pipeline/pii): split Personal-data redactors (Phone/Name/Merchant) into per-category modules |
| 4 | `c4f64525` | refactor(pipeline/pii): split Amount + HTML redactors into per-category modules                      |
| 5 | `c1e75627` | refactor(pipeline/pii): extract AuthCredentials/ErrorLog/Url/JsonBody + populate Facade              |
| 6 | `cf060369` | refactor(pipeline/pii): collapse PiiRedactor.ts to tombstone shim (30 LoC)                           |
| 7 | `ba4b5c2b` | refactor(pipeline/pii): add Section 13 max-lines cap (200 eff LoC) + canary                          |
| 8 | `079be38f` | chore(phase-6): allowlist PiiRedactor internal scaffolding + drop dead detect-dead-code entries      |
| 9 | `33d6af04` | fix(phase-6): address CodeRabbit review findings (9) on PiiRedactor split                            |

Commit 1 carries the `[phase-6-commit-1]` subject marker — the master spec §13 test-body-freeze exception that permits the two new invariant suites to land in this commit.

## Guard added in commit 7

`eslint.config.mjs` Section 13 re-imposes the file-size cap that Section 7's blanket `Types/**` exemption would otherwise hide:

- `max-lines: 200` (skipBlankLines + skipComments) on `src/Scrapers/Pipeline/Types/PiiRedactor/**/*.ts` AND on `src/Scrapers/Pipeline/EslintCanaries/no-pii-redactor-blob.canary.ts`
- The shim itself (`PiiRedactor.ts`) is intentionally left uncapped — Section 7 already allows it, and this guard is about preventing regression of the new homes, not the tombstone re-export

The 200-line ceiling is wider than §11 (NetworkDiscovery 150) / §12 (Scrape 150) because Facade and JsonBody legitimately host the cross-strategy composition logic (routing table + walker state machine) that does not split further without introducing pointless indirection.

One new canary enforces the cap (canary count: 41 → 42, all triggered):

- `no-pii-redactor-blob.canary.ts` (325 raw / 200+ eff lines — fires `max-lines`)

## Post-ship fix (commit 8)

CI `Validate` failed on the docs-coverage gate after the initial 7-commit push: 30 new top-level exports under `Types/PiiRedactor/` were flagged as introduced-but-undocumented (per-category `CATEGORY` literals, brand types, JSON-type union components, sentinel constants, Pino-censor scaffolding, AuthCredentials predicates, helpers).

All 30 are cross-module internals — the strategy-registry contract between Facade and the per-category modules. None are user-facing. Commit 8 fixes the gate in two surgical edits:

1. `.github/docs-coverage-allowlist.txt` — appends a Phase 6 section listing the 30 symbols with a `# Reason:` one-liner per entry, in the exact format established by the Phase 5 (PR #277) allowlist section. The user-facing `redact()` entry point is documented in `docs/` and so passes the gate organically (correctly NOT in the allowlist).
2. `src/Tests/Tools/detect-dead-code.ts` — removes the two Phase 6 scaffolding ENTRY_POINTS entries (`Facade.ts`, `Types.ts`) that the staged refactor used between commits 1-5. At the commit 7 final state both files are reachable via the tombstone shim's barrel re-exports; the dead-code gate stays green without the exemption (verified locally — 329 Pipeline files all have ≥1 importer).

After commit 8: `Validate` PASS (7m58s, previously FAILURE in 1m33s); `SonarCloud`, `CodeQL`, `Dependency Review`, `DNS Preflight`, `PR Title` all PASS.

## CR review cycle 1 (commit 9)

CodeRabbit auto-reviewed the PR after the draft → Ready transition (per `pr-guidlines.md §8` — no manual `@coderabbitai review` issued). Verdict: **CHANGES_REQUESTED** with 9 inline findings ([review 3514428081](https://github.com/sergienko4/israeli-bank-scrapers/pull/278#pullrequestreview-3514428081)). 8 of 9 are pre-existing behaviors preserved by the Phase 6 mechanical split; 1 (finding #9) is new in commit 1. Same fix-class precedent as Phase 5 (PR #277, `Coercion.ts` `Number('')→0`): CR findings on refactor PRs are addressed in the same PR when no public-API behavior breaks and each fix is explicitly auditable.

| # | Severity | File:Line | Pre-existing? | Resolution |
|---|----------|-----------|---------------|------------|
| 1 | 🟡 Minor    | `Amount.ts:43`           | Yes | `trim()==='' → REDACTED_HINT` |
| 2 | 🟠 Major    | `AuthCredentials.ts:65`  | Yes | `redactOtp` uses `OTP_VALUE_RE.test()` (rejects `abcd`, `12-34`) |
| 3 | 🟠 Major    | `ErrorLog.ts:50`         | Yes | Drop `isPiiRedactionDisabled` early-return — CodeQL #28 must always-redact |
| 4 | 🔵 Trivial  | `Facade.ts:157`          | Yes | Chained `endsWith` → `TOKEN_SUFFIXES`/`NAME_SUFFIXES` arrays + `.some()` |
| 5 | 🟠 Major    | `JsonBody.ts:35`         | Yes | Extract shared `LITERAL_TEXT_PII_PATTERNS` to `CommonHelpers.ts` |
| 6 | 🟠 Major    | `JsonBody.ts:78`         | Yes | Recursive `nestedHasPii`/`objectHasPii`/`arrayHasPiiObject` walker with `MAX_PII_PROBE_DEPTH=50` |
| 7 | 🛠️ Refactor | `Url.ts:99`              | Yes | Extract `listPiiQueryKeys()` helper; `redactUrl` body now 8 LoC |
| 8 | 🔴 **Critical** | `Url.ts:120`         | Yes | **PII LEAK fix** — normalize `-` separators so `4111-1111-1111-1111` collapses to `***1111` |
| 9 | 🔵 Trivial  | `DefaultDeny.test.ts:34` | No (new in commit 1) | Import `REDACTED_HINT` constant; replace 3× hardcoded `'[REDACTED]'` |

**Commit 9 diff:** 11 files changed, +166 / −75. Includes 5 new tests in `PiiRedactor.test.ts` covering findings 1, 2, 6, 8 (all tagged with `(CR review #N)` in their names for traceability) and 1 allowlist entry for the new shared `LITERAL_TEXT_PII_PATTERNS` export.

**Pre-push validation (commit 9):**
- `npx tsc --noEmit` → 0 errors
- `npx eslint "src/**/*.ts" --max-warnings=0` → 0 errors
- `npm run test:pipeline` → 4824/4824 tests pass, branches **95.03%** (above 95% threshold)
- Husky 14-gate pre-commit ladder → all gates green (no fixup-amend, single shot)

## Public surface preserved

The 30-eff-LoC `PiiRedactor.ts` tombstone shim re-exports every pre-split public symbol (no removal, no rename):

- **Identifier redactors:** `redactAccount`, `redactCard`, `redactIsraeliId`
- **Personal-data redactors:** `redactPhone`, `redactName`, `redactMerchant`
- **Amount + content redactors:** `redactAmount`, `redactHtml`, `redactUrl`, `redactUrlFull`, `redactJsonBody`
- **Credential redactors:** `redactToken`, `redactOtp`, `redactCookie`
- **Error/log helpers:** `redactErrorMessage`, `redactSensitiveEnum`
- **Pino-censor wiring:** `createCensorFn`
- **Types:** `CensorFn`, `JsonValue`, `PiiCategory`

**New additions** (allowed under the "no removal / no rename" rule; user-facing `redact()` documented in `docs/`):

- `redact(value: unknown): PiiHintString` — unified entry point that walks the strategy registry, honours always-redact for credentials, returns `REDACTED_HINT` on classifier miss (default-deny) and `REDACTION_ERROR_HINT` on strategy throw
- `REDACTED_HINT` / `REDACTION_ERROR_HINT` — `[REDACTED]` / `[REDACTION_ERROR]` literal sentinels
- `classifyKey` — promoted to the shim for cross-module symmetry with `createCensorFn`

Zero changes to caller files (`Common/ResultFormatter.ts` + ~5 other consumers + ~70 test files).

## Test safety net

`src/Tests/Unit/Pipeline/Types/PiiRedactor.test.ts` (legacy comprehensive — 96 cases), `PiiRedactor.devMode.test.ts`, and `PiiRedactorEdgeCases.test.ts` bodies are **byte-identical across commits 2-7**. Only their import paths resolve to the new per-category modules via the tombstone shim.

Commit 1 adds the only two NEW test files, both at the canonical path `src/Tests/Unit/Pipeline/Types/PiiRedactor/Invariants/`:

- `DefaultDeny.test.ts` — locks in: unclassified value → `REDACTED_HINT`
- `CredentialsNoBypass.test.ts` — locks in: `PII_REDACTION=off` does NOT bypass token / otp / cookie redaction

Both invariant suites run GREEN against the pre-split `PiiRedactor.ts` at commit 1 (per spec §6 — they document an EXISTING invariant) and GREEN against the populated Facade from commit 5 onward.

## Validation

- `npx tsc --noEmit` — clean
- `npm run lint` — clean (0 errors / 0 warnings)
- `npm run lint:dead-code` — clean (329 Pipeline files ≥1 importer)
- `npm run lint:phases:strict` — clean
- `BASE_SHA=ec40120755de182096f249dda9cc1f55083f4c55 bash .github/scripts/ci/docs-coverage.sh` — `missing: 0, allowlisted: 37, documented: 1` PASS
- All 138 `PiiRedactor`-related tests pass on every commit (5 suites: PiiRedactor, devMode, EdgeCases, DefaultDeny, CredentialsNoBypass)
- Husky pre-commit (14 gates incl. lint, type-check, tests, canaries, architecture validator, mock E2E Camoufox) green on every commit
- Canary verify: `no-pii-redactor-blob.canary.ts` fires `max-lines` error as designed (canary count 41 → 42)
- Build: `tsup` produces lib/index.cjs (854 KB) + lib/index.mjs (844 KB) + lib/index.d.ts (14 KB) without warnings

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

## Summary by CodeRabbit

* **New Features**
  * Unified redact() API and exported redaction hint tokens for consistent, visible PII markers.
  * New/redesigned redaction strategies for accounts, credentials (token/OTP/cookie), amounts, cards, URLs, HTML, names, phones, merchant IDs, Israeli IDs, JSON bodies, and error messages.

* **Tests**
  * Added invariant and edge-case tests ensuring credentials are always masked and various redaction behaviors.

* **Documentation**
  * Updated coding-quality and pre-commit guidance to enforce new linting/guideline coverage.

### CODE REVIEW

File: src/Scrapers/Pipeline/Types/PiiRedactor/Url.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Mask a single URL path segment when it looks like an identifier.
+ * @param seg - Single path segment.
+ * @returns The `***XXXX` hint when the segment looks like an ID,
+ *   otherwise the input segment unchanged.
+ */
+function maskPathSegmentIfId(seg: string): string {
+  if (!isLikelyIdSegment(seg)) return seg;
+  return redactAccount(seg);
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/Url.ts]: Addressed in 33d6af04. Added `PATH_ID_SEPARATOR_RE = /-/g` constant; `isLikelyIdSegment` and `maskPathSegmentIfId` now strip `-` separators before the digit-run check and before handing the bare digit string to `redactAccount`. Confirmed by new test `masks dash-separated card-formatted path segments (CR review #8)` — `https://x.example/api/cards/4111-1111-1111-1111/details` now collapses to `…/cards/***1111/details`. (Note on the space-separated variant: WHATWG URL parsing percent-encodes raw spaces to `%20` before path-segment splitting, so dash is the only real-world separator that reaches `isLikelyIdSegment` — kept the regex narrow + documented this in the JSDoc.)

File: src/Scrapers/Pipeline/Types/PiiRedactor/Url.ts [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+  const parse = tryParseUrl(url);
+  if (!parse.ok) return url as PiiHintString;
+  const censor = createCensorFn();
+  const allKeys = [...parse.url.searchParams.keys()];
+  const piiKeys = allKeys.filter(
+    (k): PiiClassifierBool => PII_QUERY_KEYS.has(k) as PiiClassifierBool,
+  );
+  for (const key of piiKeys) redactQueryKey(parse.url, key, censor);
+  return parse.url.toString() as PiiHintString;
+}
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/Url.ts:108]: Addressed in 33d6af04. Extracted `listPiiQueryKeys(parsed: URL)` helper. `redactUrl` body is now 8 LoC (well under CLAUDE.md `Max 10 lines per method` cap). Verified by manual line count and the project `max-lines-per-function: 10` ESLint rule (now passes on this file).

File: src/Scrapers/Pipeline/Types/PiiRedactor/Amount.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * Amount strategy. Returns sign-only marker.
+ * @param value - Number or numeric string.
+ * @returns Stable hint.
+ */
+function redact(value: number | string): PiiHintString {
+  if (isPiiRedactionDisabled) return String(value) as PiiHintString;
+  const num = coerceToNumber(value);
+  if (Number.isNaN(num)) return REDACTED_HINT as PiiHintString;
+  if (num < 0) return '-***' as PiiHintString;
+  return '+***' as PiiHintString;
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/Amount.ts]: Addressed in 33d6af04. `redactAmount('')` and `redactAmount('   ')` now return `REDACTED_HINT` instead of `+***`. Confirmed by 2 new tests in `PiiRedactor.test.ts` (`returns [REDACTED] for empty string (CR review #1)` and `returns [REDACTED] for whitespace-only string (CR review #1)`).

File: src/Tests/Unit/Pipeline/Types/PiiRedactor/Invariants/DefaultDeny.test.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    const empty = {};
+    const result = redact(empty);
+    expect(result).toBe('[REDACTED]');
+  });
+
+  it('returns the REDACTED hint for unclassified plain strings', () => {
+    const debugString = 'some-arbitrary-debug-string';
+    const result = redact(debugString);
+    expect(result).toBe('[REDACTED]');
+  });
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Tests/Unit/Pipeline/Types/PiiRedactor/Invariants/DefaultDeny.test.ts]: Addressed in 33d6af04. The import now includes `REDACTED_HINT` from the public `PiiRedactor` entry point, and all 3 `expect(result).toBe('[REDACTED]')` calls are replaced with `toBe(REDACTED_HINT)`. The hint literal is no longer duplicated between production and test.

File: src/Scrapers/Pipeline/Types/PiiRedactor/JsonBody.ts [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+
+/** Maximum walk depth before redactJsonBody bails out for safety. */
+const MAX_WALK_DEPTH = 1000;
+
+/** Regex set used by the JSON-fallback path when the body isn't valid JSON. */
+const FALLBACK_PATTERNS: readonly { readonly re: RegExp; readonly to: string }[] = [
+  { re: /\b(\d{2}-\d{3}-)\d+(\d{4})\b/g, to: '$1***$2' },
+  { re: /(?<!\d)\d{5}(\d{4})(?!\d)/g, to: '***$1' },
+  { re: /eyJ[\w-]{20,}/g, to: REDACTED_HINT },
+];
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/JsonBody.ts]: Addressed in 33d6af04. The 3-regex fallback table is now exported from `CommonHelpers.ts` as `LITERAL_TEXT_PII_PATTERNS` and consumed by both `JsonBody.applyFallbackPatterns` and `Html.redactHtml`. Local duplicates in both files are gone. The new export is allowlisted in `.github/docs-coverage-allowlist.txt` with a `# Reason:` comment per Phase-5 precedent. Naming intentionally diverges from `FALLBACK_PATTERNS` because the table is no longer a 'fallback' in the HTML context — it's the primary scrubbing pass for free-form text nodes.

File: src/Scrapers/Pipeline/Types/PiiRedactor/Facade.ts [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+ * @returns True when the key looks like a personal-name field.
+ */
+function isNameSuffix(key: string): PiiClassifierBool {
+  const lower = key.toLowerCase();
+  if (lower.endsWith('firstname')) return true as PiiClassifierBool;
+  if (lower.endsWith('lastname')) return true as PiiClassifierBool;
+  if (lower.endsWith('fullname')) return true as PiiClassifierBool;
+  if (lower.endsWith('customername')) return true as PiiClassifierBool;
+  return false as PiiClassifierBool;
+}
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/Facade.ts:155]: Addressed in 33d6af04. `classifyKey` now uses two `as const` config arrays (`TOKEN_SUFFIXES` for token/bearer/cookie/secret, `NAME_SUFFIXES` for firstname/lastname/fullname/customername) with `.some(s => lower.endsWith(s))`. The chained `if` branches are gone, satisfying CLAUDE.md SOLID/OCP rule (`maps over if/else`).

File: src/Scrapers/Pipeline/Types/PiiRedactor/JsonBody.ts [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+/**
+ * Whether an array contains at least one plain-object element with PII.
+ * @param arr - Candidate array.
+ * @returns True when at least one element has a PII-classified key.
+ */
+function arrayHasPiiObject(arr: JsonArray): PiiClassifierBool {
+  return arr.some((el): PiiClassifierBool => {
+    if (!isJsonObject(el)) return false as PiiClassifierBool;
+    return objectHasPii(el);
+  }) as PiiClassifierBool;
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/JsonBody.ts]: Addressed in 33d6af04. Added `nestedHasPii` + `objectHasPii` + `arrayHasPiiObject` recursive trio with a `MAX_PII_PROBE_DEPTH = 50` defensive bound. `arrayHasPiiObject` now correctly flags arrays of plain objects with nested PII (`[{ meta: { phoneNumber: '…' } }]` collapses to `[<N redacted items>]`). Confirmed by 2 new tests: `collapses arrays whose elements carry nested PII (CR review #6)` and `detects PII nested inside arrays-of-arrays-of-objects`. ESLint `max-depth=1` and TS strict pass.

File: src/Scrapers/Pipeline/Types/PiiRedactor/ErrorLog.ts [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+ * values with the stable token `<REDACTED_ENUM>`; non-sensitive values
+ * pass through unchanged.
+ * @param value - Raw scraper-error-type enum value.
+ * @returns `<REDACTED_ENUM>` for sensitive values, otherwise input.
+ */
+function redactSensitiveEnum(value: string): PiiHintString {
+  if (isPiiRedactionDisabled) return value as PiiHintString;
+  if (value.length === 0) return value as PiiHintString;
+  if (SENSITIVE_SCRAPER_ENUMS.has(value)) return '<REDACTED_ENUM>' as PiiHintString;
+  return value as PiiHintString;
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/ErrorLog.ts:52]: Addressed in 33d6af04. The `isPiiRedactionDisabled` early-return has been removed from BOTH `redactErrorMessage` and `redactSensitiveEnum`. The function-level JSDoc now explicitly cites CodeQL js/clear-text-logging alert #28 as the rationale for always-redacting these payloads regardless of the dev-mode bypass. Verified `ResultFormatter` test suite (the primary consumer) still passes (168/168 tests across the 6-suite `PiiRedactor|ResultFormatter|ErrorLog` slice).

File: src/Scrapers/Pipeline/Types/PiiRedactor/AuthCredentials.ts [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+ * OTP strategy. Returns '[OTP]' for 4..8 digit inputs; default-deny otherwise.
+ * @param value - Raw OTP.
+ * @returns Stable hint.
+ */
+function redactOtp(value: string): PiiHintString {
+  if (isPiiRedactionDisabled) return value as PiiHintString;
+  if (value.length === 0) return '' as PiiHintString;
+  if (value.length < OTP_MIN_LEN) return REDACTED_HINT as PiiHintString;
+  if (value.length > OTP_MAX_LEN) return REDACTED_HINT as PiiHintString;
+  return '[OTP]' as PiiHintString;
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply src/Scrapers/Pipeline/Types/PiiRedactor/AuthCredentials.ts]: Addressed in 33d6af04. `redactOtp` now uses `OTP_VALUE_RE.test(value)` (the regex we already used for shape sniffing) so length-OK strings with non-digit characters fall through to `REDACTED_HINT`. Dropped the now-redundant `OTP_MIN_LEN`/`OTP_MAX_LEN` constants. Confirmed by 2 new tests (`returns [REDACTED] for length-OK but non-digit string` and `returns [REDACTED] for length-OK mixed digits + separators`).

File: docs/contributing/lint.md [hunk 10]
```diff
@@ -1,6 +1,8 @@
 # Code style & lint
 
-The pre-commit hook runs ESLint + Biome + Prettier + the architecture validator + canaries on every commit. Skipping any of them is not an option.
+The pre-commit hook runs ESLint + Biome + Prettier + the architecture validator + canaries + guideline-coverage on every commit. Skipping any of them is not an option.
+
+> **Quality caps** (per-function LoC, file size, complexity, parameter count) live in [CLEAN_CODE.md](../../CLEAN_CODE.md). That file is the single source of truth — do NOT restate the numbers here. The `lint:guideline-coverage` gate asserts `eslint.config.mjs` actually enforces those caps for every Pipeline cluster.
```
[the remark this answers is not in the document]
Comment (sergienko4) [reply docs/contributing/lint.md]: Addressed in e373b6cd. Switched to the `{{BRANCH}}` placeholder pattern (matches the existing usage at lint.md:65 used by the mkdocs branch-substitution plugin). Verified locally with `python -m mkdocs build --strict` (now exits 0; previously aborted on this exact link).

Also added a new Husky pre-commit gate `docs-strict` (gate 11/14) that runs `mkdocs build --strict` whenever `docs/**`, root `*.md`, or `mkdocs.yml` is staged — would have caught this locally. Soft-skips when Python/mkdocs not on PATH so cross-platform contributors aren't blocked (CI is still the hard gate).
